discovery/utils/visualizer.py

```diff
@@ -100,29 +121,34 @@ def _generate_table(self, filename: str, col_strings: str):
               {col_strings}
             </TABLE>>'''
 
-    def _determine_working_node(self, full_path):
-        # directory already registered as a node
-        already_visited = next(iter([x for x in self.visited_nodes if x.fs_full_path == full_path]), None)
-        if already_visited is not None:
-            return already_visited
+    def _determine_working_node(self, filesystem_full_path: str):
+        determined_node = None
 
-        new_node = SubGraphNode(full_path)
+        already_visited_node = self._get_node_if_path_was_already_observed(filesystem_full_path)
+        if already_visited_node is not None:
+            determined_node = already_visited_node
+        else:
+            determined_node = SubGraphNode(filesystem_full_path)
+            self._add_all_previously_unobserved_parents_to_observed_nodes(determined_node)
+        return determined_node
 
-        # ensure that the tree has all ancenstors, even if no files were registered previously
-        child = new_node
+    def _add_all_previously_unobserved_parents_to_observed_nodes(self, parameter_child_node: SubGraphNode):
+        working_child_node = parameter_child_node
         while True:
-            visited_parent = self._get_visited_parent(child)
+            visited_parent = self._get_parent_if_parent_was_already_observed(working_child_node)
             if visited_parent is None:
-                new_parent = SubGraphNode(child.get_parent_path())
-                new_parent.children.append(child)
-
+                new_parent = SubGraphNode(working_child_node.get_parent_path())
+                new_parent.children.append(working_child_node)
                 # setup next loop
-                child = new_parent
+                working_child_node = new_parent
             else:
-                visited_parent.children.append(child)
+                visited_parent.children.append(working_child_node)
                 break
-        return new_node
 
-    def _get_visited_parent(self, node: SubGraphNode):
-        return next(iter([x for x in self.visited_nodes
-                          if x.fs_full_path == node.get_parent_path()]), None)
+    def _get_node_if_path_was_already_observed(self, filesystem_full_path: str):
```

Comment:
These two methods should probably not exist, if they must, you can abstract out the 
`list[0] if list else None` functionality into a class or function

Reply:
I don't understand why they shouldn't exist, perhaps you misunderstood something? These functions are already very simple abstractions?

Reply:
My idea was that they violate the DRY principle, since their only purpose is to access a single value that matches
I think the best solution would be to rebuild the types as dicts, so they can be index'd, which I think would remove a lot of confusion as to what you're trying to acces